src/clients/UBAClient/UBAClientAbstract.ts

```diff
@@ -0,0 +1,139 @@
+   * @param spokePoolToken The token to get the opening balance for
+   * @param hubPoolBlockNumber The block number to get the opening balance for
+   * @returns The opening balance for the given token on the given chainId at the given block number
+   * @throws If the token cannot be found for the given chainId
+   * @throws If the opening balance cannot be found for the given token on the given chainId at the given block number
+   */
+  public abstract getOpeningBalance(
+    chainId: number,
+    spokePoolToken: string,
+    hubPoolBlockNumber?: number
+  ): OpeningBalanceReturnType;
+
+  /**
+   * Gets the latest block number for a given chainId in the state of the lastest closing block
+   * @param chainId The chainId to get the latest block number for
+   * @returns The latest block number for the given chainId
+   * @note Assumes that the `spoke[...].bundleEndBlocks` are sorted in ascending order
+   */
+  protected abstract resolveClosingBlockNumber(chainId: number, blockNumber: number): number;
+
+  /**
+   * @description Construct the ordered sequence of SpokePool flows between two blocks.
+   * @note Assumptions:
+   * @note Deposits, Fills and RefundRequests have been pre-verified by the SpokePool contract or SpokePoolClient, i.e.:
+   * @note - Deposit events contain valid information.
+   * @note - Fill events correspond to valid deposits.
+   * @note - RefundRequest events correspond to valid fills.
+   * @note In order to provide up-to-date prices, UBA functionality may want to follow close to "latest" and so may still
+   * @note be exposed to finality risk. Additional verification that can only be performed within the UBA context:
+   * @note - Only the first instance of a partial fill for a deposit is accepted. The total deposit amount is taken, and
+   * @note   subsequent partial, complete or slow fills are disregarded.
+   * @param spokePoolClient SpokePoolClient instance for this chain.
+   * @param fromBlock       Optional lower bound of the search range. Defaults to the SpokePool deployment block.
+   * @param toBlock         Optional upper bound of the search range. Defaults to the latest queried block.
+   */
+  public abstract getFlows(chainId: number, fromBlock?: number, toBlock?: number): UbaFlow[];
+
+  /**
+   * @description Evaluate an RefundRequest object for validity.
+   * @dev  Callers should evaluate 'valid' before 'reason' in the return object.
+   * @dev  The following RefundRequest attributes are not evaluated for validity and should be checked separately:
+   * @dev  - previousIdenticalRequests
+   * @dev  - Age of blockNumber (i.e. according to SpokePool finality)
+   * @param chainId       ChainId of SpokePool where refundRequest originated.
+   * @param refundRequest RefundRequest object to be evaluated for validity.
+   */
+  public abstract refundRequestIsValid(
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _chainId: number,
```

Comment:
FWIW the reason for including the chainId here is because the RefundRequest[WithBlock] type doesn't actually identify which chain the transaction occurred on. This is more or less unique amongst our SpokePool events; we might want to consider adding something like `refundChainId` to the RefundRequest type.

Reply:
Since this is just an abstract fn, the naming of the variables is mostly optional. However, the `_` was a carryover from my initial implementations. Thanks for catching this 🙂 https://github.com/across-protocol/sdk-v2/pull/236/commits/c881c1bb60c36b62cc9b342dc718b52e4541bd6c